options/options.js

```diff
@@ -0,0 +1,141 @@
+(() => {
+  const profilesWrapper = document.getElementById('profiles-wrapper');
+  const addButton = document.getElementById('add');
+  const PROFILES = 'profiles';
+
+  const uuid = () => {
+    return crypto.randomUUID();
+  }
+
+  const getProfiles = async () => {
+    return (await browser.storage.sync.get(PROFILES))[PROFILES] || [];
+  };
+
+  const saveProfile = async (profile) => {
+    const storedProfiles = await getProfiles();
+    const existingProfile = storedProfiles.find(sp => sp.id === profile.id);
+    let updatedProfiles;
+
+    if (existingProfile) {
+      existingProfile.content = profile.content;
+      existingProfile.name = profile.name;
+
+      updatedProfiles = {
+        [PROFILES]: storedProfiles,
+      };
+    } else {
+      updatedProfiles = {
+        [PROFILES]: [
+          ...storedProfiles,
+          profile,
+        ]
+      };
+    }
+
+    return browser.storage.sync.set(updatedProfiles);
+  };
+
+  const deleteProfile = async (id) => {
+    const storedProfiles = await getProfiles();
+    const updatedProfiles = storedProfiles.filter(sp => sp.id !== id);
+
+    return browser.storage.sync.set({ [PROFILES]: updatedProfiles });
+  };
+
+  const onSaveProfile = (event) => {
+    const targetProfile = event.target.parentElement.parentElement;
+    const name = targetProfile.children[0].value;
+    const content = targetProfile.children[1].value
+      .trim().split('\n')
+      .map(line => line.trim());
+
+    if (!name) {
+      return;
+    }
+
+    const backgroundPage = browser.extension.getBackgroundPage();
+    let profileFunction = 'updateProfile';
+
+    // Save new profile
+    if (!targetProfile.dataset.id) {
+      targetProfile.dataset.id = uuid();
+      targetProfile.classList.remove('new-profile');
+      profileFunction = 'createProfile';
+    }
+
+    const id = targetProfile.dataset.id;
+    const profile = { id, name, content };
+
+    backgroundPage[profileFunction](profile);
```

Comment:
It can be either `createProfile` or `updateProfile`. I used the indirection to avoid duplicating that part which shares most of the logic (updateProfile is missing the id creation and removal of a css class).

This is a valid alternative. Let me know if you prefer the change or have another suggestion.
```javascript
    // Save new profile
    if (!targetProfile.dataset.id) {
      // Unique to createProfile
      targetProfile.dataset.id = crypto.randomUUID();
      targetProfile.classList.remove('new-profile');

      // Shared with updateProfile
      const id = targetProfile.dataset.id;
      backgroundPage.createProfile({ id, name, content });
 
     return saveProfile(profile);
    }

    const id = targetProfile.dataset.id;
    backgroundPage.updateProfile({ id, name, content });

    return saveProfile(profile);
```

Reply:
That's a bit clearer to me, thanks!

Reply:
I think this is the last remaining thing; one we "unroll" the control flow here, this PR is good to land.